File: Java/medium/421. Maximum XOR of Two Numbers in an Array/421_Maximum_XOR_of_Two_Numbers_in_an_Array.java
public class Solution {
    public int findMaximumXOR(int[] nums) {
        int max = 0, mask = 0;
        for (int i = 31; i >= 0; i--) {
            mask |= (1 << i);
            Set<Integer> prefixes = new HashSet<>();
            for (int num : nums) {
                prefixes.add(num & mask);
            }
            int candidate = max | (1 << i);
            for (int prefix : prefixes) {
                if (prefixes.contains(prefix ^ candidate)) {
                    max = candidate;
                    break;
                }
            }
        }
        return max;
    }
}

File: Golang/421. Maximum XOR of Two Numbers in an Array/421_Maximum_XOR_of_Two_Numbers_in_an_Array.go
func findMaximumXOR(nums []int) int {
	maxXOR, mask := 0, 0
	for i := 31; i >= 0; i-- {
		mask |= 1 << i
		prefixSet := map[int]bool{}
		for _, num := range nums {
			prefixSet[num&mask] = true
		}
		candidate := maxXOR | (1 << i)
		for prefix := range prefixSet {
			if prefixSet[prefix^candidate] {
				maxXOR = candidate
				break
			}
		}
	}
	return maxXOR
}
